### PR TITLE
mbox: initial import of a thread-independent IPC

### PR DESCRIPTION
**Pull request by authmillenon**
Introduction
=========
I know @kaspar030 is working on something like this too, but when I was working with lwIP again, I decided to port out the messaging layer I implemented for it, to better test it and to give it as an alternative to @kaspar030's work (and since it was already laying around in my repo it took me only half a day and some hours of boredom on Christmas to port it to `msg_t`).

`mbox` is thought to be an alternative - not a replacement - to standard-IPC. There are probably many ways to make it a little bit smaller in memory consumption, but I based it on the [unix support for mbox of lwIP](http://git.savannah.gnu.org/cgit/lwip/lwip-contrib.git/tree/ports/unix/sys_arch.c) (with many adaptations to make it more RIOT-y) so it is quite stable.

Summary
=======
`mbox` is a datatype with two base operations (of that there are some variations - more on that later):
 - `mbox_put()` and
 - `mbox_get()`

Blocking IPC
-----------------
`mbox_put()` allows a user to put a `msg_t` in the `mbox`. If the `mbox` is full the thread will block and be signaled through the `mbox::not_full` semaphore, when there is space in the `mbox` again.

`mbox_get()` allows a user to get a `msg_t` from the `mbox`. If the `mbox` is empty the thread will block and be signaled through the `mbox::not_empty` semaphore, when another thread puts something in the `mbox`.

Non-blocking IPC
-----------------------
There is a non-blocking variant `mbox_tryget()` and `mbox_tryput()` for both operations, that just returns `-EAGAIN` when the normal operations would block (they don't require the semaphores, though they both post them, so the blocking variants are woken up).

mbox and time
-------------------
Since the blocking mechanism is based on semaphores there is also a timeout-variant of `mbox_get()`: `mbox_get_timed()`. If we make this optional for semaphores we also can make it optional here.

Dependencies
==========
Depends on #4551 and maybe #4374 (though I did not run into any problems during tests, regarding this bug).

### CODE REVIEW

**Comment by OlegHahm** [comment]
Sorry, I don't have time to review this.

**Comment by authmillenon** [comment]
@Kijewski since you have some experience with semaphores also, you might want to have a look too.

**Comment by kaspar030** [comment]
While I like the simplicity of the approach, it doesn't tackle some of the IMHO important aspects:

- it cannot be used from within an ISR
- it's kinda fat memory-wise (sizeof(mbox_t) = 44)
- it's kinda slow (about a third of the speed of our current messaging)
- it doesn't allow waiting on multiple queues
(- it doesn't allow arbitrarily sized messages, but that's maybe not needed)

**Comment by authmillenon** [comment]
> it cannot be used from within an ISR

Could be fixed easily IMHO

> it's kinda fat memory-wise (sizeof(mbox_t) = 44)

As I mentioned in the description: there is room for optimization ;) 

> it's kinda slow (about a third of the speed of our current messaging)

Since semaphores are based on the current messaging this doesn't surprise me really

> it doesn't allow waiting on multiple queues

Was that something we wanted? This should be adaptable easily, but it wouldn't make it stronger in the points you already gave. 

> (it doesn't allow arbitrarily sized messages, but that's maybe not needed)

What do you mean by that?

**Comment by kaspar030** [comment]
@authmillenon Well, devising the API (mbox_put(), mbox_get(), ...) is simple, the other points make it hard ;)

>> (it doesn't allow arbitrarily sized messages, but that's maybe not needed)
> 
> What do you mean by that?

e.g.
```
mbox_init(mbox, num, size);
mbox_put(mbox, obj, size);
```

**Comment by authmillenon** [comment]
Except for the fixed size for the message box this is exactly how it was done in lwIP (with void pointers). I'm not sure however, how this could put to a better use, than a fixed `msg_t` already can.

**Comment by OlegHahm** [comment]
Needs a rebase, but is no longer waiting on anything.

**Comment by authmillenon** [comment]
Rebased

File: sys/include/mbox.h [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+ * @param[in] array Pointer to preallocated array of @ref msg_t structures.
+ * @param[in] num   Number of @p msg_t structures in @p array.
+ *
+ * @return  Statically initialized message box
+ */
+#define MBOX_CREATE(array, num) { array, CIB_INIT(num), MUTEX_INIT, \
+                                  SEMA_CREATE(1), SEMA_CREATE(0), 0 }
+
+/**
+ * @brief   Static initializer for message box.
```
Comment (OlegHahm) [inline sys/include/mbox.h]: Static?

File: sys/include/mbox.h [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+ * This function will not block if @p mbox is full, but signal this fact through its return value.
+ *
+ * @param[in] mbox  The message box to put the message into.
+ * @param[in] msg   A message.
+ *
+ * @return  0, on success.
+ * @return  -EAGAIN, if @p mbox was full.
+ * @return  -EOVERFLOW, if one of the mbox::not_empty would overflow.
+ */
+int mbox_tryput(mbox_t *mbox, msg_t *msg);
```
Comment (OlegHahm) [inline sys/include/mbox.h]: For consistency: `mbox_try_put()`

**Comment by kaspar030** [comment]
I have #4919 lying around, please consider that.

**Comment by authmillenon** [comment]
I still need to test #4919, but I already know that I will prefer it => close.

**Pull request closed without merging**